### ISSUE TITLE
Decoding values greater than 2^31 - 1 will cause an integer overflow.

### ISSUE THREAD

**Issue by leonardolarrad** [bolderflight/uleb128#2]
Hi!

The following test fails due to an integer overflow. This overflow consistently occurs when decoding values greater than 2^31 - 1. 

```cpp
/* Test input > 2^31 - 1 */
TEST(Uleb128, RealMidInput) {
  uint8_t buf[10];
  uint64_t input = 2147483648ull; // 2^31
  uint64_t output = 0;
  std::size_t bytes_written = bfs::EncodeUleb128(input, buf, sizeof(buf));
  std::size_t bytes_read = bfs::DecodeUleb128(buf, sizeof(buf), &output); // !! integer overflow
  EXPECT_EQ(5, bytes_written);
  EXPECT_EQ(5, bytes_read);
  EXPECT_EQ(input, output); // !! failed

  //Expected equality of these values:
  // input
  //   Which is: 2147483648
  // output
  //   Which is: 18446744071562067968
}
```

The test `MaxInput` did not catch this behavior because the overflow sets the output value exactly to uint64_t maximum, giving the impression that the condition `EXPECT_EQ(std::numeric_limits<uint64_t>::max(), output);` is met.

```cpp
std::size_t DecodeUleb128(uint8_t const * const data, const std::size_t len,
                          uint64_t * const val) {
  /* Null pointer check */
  if ((!data) || (!val)) {return 0;}
  uint64_t res = 0, shift = 0;
  std::size_t i = 0;
  while (1) {
    /* Prevent buffer overflow */
    if (i < len) {
      uint8_t b = data[i++];
      res |= (b & 0x7F) << shift;   // uleb128.h:67: !! integer overflow
      if (!(b & 0x80)) {break;}
      shift += 7;
    } else {
      return 0;
    }
  }
  *val = res;
  return i;
}
```

Replacing `uleb128.h:67` with the following solves the issue:

```cpp
res |= uint64_t(b & 0x7F) << shift;
```

**Comment by flybrianfly** [comment]
Thanks for catching that! Makes sense, will incorporate that fix and the additional test.

**Comment by flybrianfly** [comment]
Fixed in v2.1.2
